issue_6/contracts/GreenAMM.sol

```diff
@@ -0,0 +1,360 @@
+    @param amountX X tokens to deposit
+    @param amountY Y tokens to deposit
+    @return rewardsFraction how many AMM tokens need to be issued (36-decimal fraction of total supply)
+     */
+    function stateChangeMint(uint256 amountX, uint256 amountY)
+        internal
+        initialized
+        maintainsBalance(amountX, amountY)
+        returns (uint256 rewardsFraction)
+    {
+        rewardsFraction = FPMath.wadDiv(amountX, x.reserves);
+
+        x.reserves += amountX;
+        y.reserves += amountY;
+
+        kRecompute();
+    }
+
+    /**
+    @notice 'State Chage' function for 'Burn' method
+    @param amountX X tokens to burn
+    @param amountY Y tokens to burn
+    @return rewardsFraction how many AMM tokens need to be burned (36-decimal fraction of total supply)
+     */
+    function stateChangeBurn(uint256 amountX, uint256 amountY)
+        internal
+        initialized
+        maintainsBalance(amountX, amountY)
+        returns (uint256 rewardsFraction)
+    {
+        rewardsFraction = FPMath.wadDiv(amountX, x.reserves);
+
+        x.reserves -= amountX;
+        y.reserves -= amountY;
+
+        kRecompute();
+
+        // BurnInteraction(amountX, amountY, rewardsAmount);
+    }
+
+    /**
+    @notice 'State Chage' function for 'SellX' method
+    @param amountX X tokens to add to reserves
+    @return how many Y tokens need to be returned to the user
+     */
+    function stateChangeSellX(uint256 amountX) internal initialized returns (uint256) {
+        // (x + A) * (y - B) = k
+        // y - B = k / (x + A)
+        // B = y - k / (x + A)
+        uint256 amountY = y.reserves - k / (x.reserves + amountX);
+        x.reserves += amountX;
+        y.reserves -= amountY;
+
+        return amountY;
+    }
+
+    /**
+    @notice 'State Chage' function for 'SellY' method
+    @param amountY Y tokens to add to reserves
+    @return how many X tokens need to be returned to the user
+     */
+
+    function stateChangeSellY(uint256 amountY) internal initialized returns (uint256) {
+        uint256 amountX = x.reserves - k / (y.reserves + amountY);
+        x.reserves -= amountX;
+        y.reserves += amountY;
+
+        return amountX;
+    }
+}
+
+/**
+@notice Standard k=xy AMM contract
+ */
+contract GreenAMM is ERC20("GreenAMMToken", "OWL_GAMT", 36), GreenAMMEngine {
+    /// @dev X token
+    IERC20Metadata internal xToken;
+    /// @dev Y token
+    IERC20Metadata internal yToken;
+
+    constructor(IERC20Metadata _x, IERC20Metadata _y) GreenAMMEngine(_x.decimals(), _y.decimals()) {
+        require(address(_x) != address(_y), "bad pair");
+        xToken = _x;
+        yToken = _y;
+    }
+
+    /**
+    @notice Supply any amount of x and y to the contract
+    Mint k AMM tokens to sender. Can only be called once.
+
+    @param amountX X tokens to deposit
+    @param amountY Y tokens to deposit
+    */
+    function Init(uint256 amountX, uint256 amountY) public {
```

Comment:
Functions are in lowerCamelCase, contracts, structs and events are in UpperCamelCase